thrifty-runtime/src/main/java/com/microsoft/thrifty/service/ClientBase.java

```diff
@@ -85,226 +48,96 @@
     /**
      * A flag indicating whether the client is active and connected.
      */
-    private final AtomicBoolean running = new AtomicBoolean(true);
-
-    /**
-     * A single-thread executor on which to invoke method callbacks.
-     *
-     * <p>I expect that we'll revisit this design choice; it guarantees
-     * that method responses won't race each other, but arguably that's
-     * a higher-level concern, and this does feel a bit heavy-handed.
-     */
-    private final ExecutorService callbackExecutor = Executors.newSingleThreadExecutor();
-
-    /**
-     * An unbounded queue holding RPC calls awaiting execution.
-     */
-    private final BlockingQueue<MethodCall<?>> pendingCalls = new LinkedBlockingQueue<>();
+    final AtomicBoolean running = new AtomicBoolean(true);
 
     private final Protocol protocol;
-    private final Listener listener;
-    private final WorkerThread workerThread;
 
-    protected ClientBase(Protocol protocol, Listener listener) {
+    protected ClientBase(Protocol protocol) {
         this.protocol = protocol;
-        this.listener = listener;
-        this.workerThread = new WorkerThread();
-
-        workerThread.setDaemon(true);
-        workerThread.start();
     }
 
     /**
-     * When invoked by a derived instance, places the given call in a queue to
-     * be sent to the server.
+     * When invoked by a derived instance, sends the given call to the server.
      *
      * @param methodCall the remote method call to be invoked
+     * @return the result of the method call
      */
-    protected void enqueue(MethodCall<?> methodCall) {
+    protected final Object execute(MethodCall<?> methodCall) throws Exception {
```

Comment:
Do we want this here or a second `ClientBase` subclass like `SyncClientBase` (in that case `handleExceptionMessage()` could be abstract in here).